### PR TITLE
Store built image digest in the build status

### PR DESCRIPTION
**Pull request by mmilata**
Trello card: https://trello.com/c/SAvLJLLl/1036-5-provide-the-built-image-reference-as-part-of-the-build-status-builds

@legionus @miminar I'm extracting the digest from the stream of jsons docker client generates during pushing, is this approach fine with you guys? I've asked Docker if they consider it part of the API: docker/docker#29865.

@bparees PTAL

### CODE REVIEW

**Comment by mmilata** [comment]
[test][testextended][extended:core(builds)]

**Comment by bparees** [comment]
lgtm, @openshift/api-review please approve the new digest field.

File: pkg/build/api/types.go [hunk 1]
```diff
@@ -218,6 +218,11 @@ type BuildStatus struct {
 
 	// Config is an ObjectReference to the BuildConfig this Build is based on.
 	Config *kapi.ObjectReference
+
+	// outputDockerImageDigest is the digest of the built Docker image. The
+	// digest uniquely identifies the image in the registry to which it was
+	// pushed.
+	OutputDockerImageDigest string
```
Comment (smarterclayton) [inline pkg/build/api/types.go]: Put this in a struct that parallels spec output - probably `status.output.to.imageDigest` where output is a struct, to is a pointer to a struct, and to is only set when an image was output.

**Comment by smarterclayton** [comment]
Also add at to the field description that the field may not be set the push
completes but the build encounters an unexpected failure.  That is, the
field is set "best effort".

File: pkg/build/api/types.go (hunk 1 above)
Comment (mmilata) [inline pkg/build/api/types.go]: I've changed the structs as requested, PTAL if it's what you wanted.

**Comment by mmilata** [comment]
Rebased, though there's still a bug in the digest extraction code that makes the tests fail that I'm working on.

File: pkg/build/builder/dockerutil_test.go [hunk 2]
```diff
@@ -168,8 +175,267 @@ func TestPushImage(t *testing.T) {
...
+			},
+		},
+	}
+
+main:
+	for i, tc := range tests {
+		decoded := []progressLine{}
+		w := newPushWriter(func(line progressLine) error {
+			decoded = append(decoded, line)
+			if line.Error != "" {
```
Comment (miminar) [inline pkg/build/builder/dockerutil_test.go]: `len(line.Error) > 0` and the same below

File: test/extended/builds/digest.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+				g.By("creating test build")
+				err = oc.Run("create").Args("-f", stiBuildFixture).Execute()
+				o.Expect(err).NotTo(o.HaveOccurred())
+
+				g.By("starting a test build")
+				br, err := exutil.StartBuildAndWait(oc, "test")
+
+				g.By("checking that the resulting image has a digest")
+				o.Expect(br.Build.Status.Output.To.ImageDigest).To(o.HavePrefix("sha256:"))
```
Comment (miminar) [inline test/extended/builds/digest.go]: Can you also fetch the created image object by istag and match its name against the digest?

File: test/extended/builds/digest.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+			})
+		})
+
+		g.Describe("started with log level >5", func() {
+			g.It(fmt.Sprintf("should save the image digest when finished"), func() {
```
Comment (miminar) [inline test/extended/builds/digest.go]: All the `It` blocks differ in just few values. Could you make it a top-level parametrised function and call it from each sub-Describe?

File: pkg/build/builder/docker.go [hunk 5]
```diff
@@ -173,14 +162,19 @@ func (d *DockerBuilder) Build() error {
...
+		if err != nil {
 			d.build.Status.Reason = api.StatusReasonPushImageToRegistryFailed
 			d.build.Status.Message = api.StatusMessagePushImageToRegistryFailed
-			if updateErr := retryBuildStatusUpdate(d.build, d.client, nil); updateErr != nil {
-				utilruntime.HandleError(fmt.Errorf("error: An error occured while updating the build status: %v", updateErr))
-			}
+			handleBuildStatusUpdate(d.build, d.client, nil)
 			return reportPushFailure(err, authPresent, pushAuthConfig)
 		}
+		if digest != "" {
```
Comment (mmilata) [inline pkg/build/builder/docker.go]: Just curious, is it a matter of consistency or is there a practical difference?

File: test/extended/builds/digest.go (hunk 4 above)
Comment (mmilata) [inline test/extended/builds/digest.go]: Huh, I even searched the ginkgo docs to see if it supports parametrized tests, missing the obvious solution completely /facepalm.

File: pkg/build/builder/docker.go (hunk 5 above)
Comment (miminar) [inline pkg/build/builder/docker.go]: I'd say it's just about consistency.

**Comment by mmilata** [comment]
@miminar thank you for the review. I've addressed the issues you found and pushed updated branch.

**Comment by bparees** [comment]
@openshift/api-review needs api approval to land for 3.5.

**Comment by mmilata** [comment]
Flake #12341, #12479.

File: pkg/build/builder/docker.go (hunk 5 above)
Comment (miminar) [inline pkg/build/builder/docker.go]: bump :-)

File: pkg/build/builder/sti.go [hunk 6]
```diff
@@ -305,14 +293,19 @@ func (s *S2IBuilder) Build() error {
...
+		if err != nil {
 			s.build.Status.Reason = api.StatusReasonPushImageToRegistryFailed
 			s.build.Status.Message = api.StatusMessagePushImageToRegistryFailed
-			if updateErr := retryBuildStatusUpdate(s.build, s.client, nil); updateErr != nil {
-				utilruntime.HandleError(fmt.Errorf("error occured while updating the build status: %v", updateErr))
-			}
+			handleBuildStatusUpdate(s.build, s.client, nil)
 			return reportPushFailure(err, authPresent, pushAuthConfig)
 		}
+		if digest != "" {
```
Comment (miminar) [inline pkg/build/builder/sti.go]: `len()`

Comment (mmilata) [inline pkg/build/builder/sti.go]: Fixed, thanks:)
```
git show | grep -c '^[+-].*[!=]= ""'      
0
```

**Comment by mmilata** [comment]
Flake #12479. [testextended][extended:core(builds)]

**Comment by smarterclayton** [comment]
API approved.

**Comment by bparees** [comment]
extended test failure is a known flake but it's also in an area of code related to what's being changed here, so i'd like to see a clean run.

[testextended][extended:core(failure reason)]

**Comment by bparees** [comment]
seems like a consistent failure that needs to be investigated.

**Comment by mmilata** [comment]
Flake #9414. Trying to reproduce the failing extended tests locally.

**Comment by bparees** [comment]
flake https://github.com/openshift/origin/issues/9414
[merge]

**Pull request merged**